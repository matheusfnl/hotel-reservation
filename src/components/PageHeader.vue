<template>
  <q-card class="bg-white rounded-borders-md">
    <q-card-section class="q-px-md q-py-sm">
      <h2 class="text-h5 q-ma-none">{{ $t('hotel.title') }}</h2>
    </q-card-section>

    <q-separator />

    <q-card-section class="column q-gutter-md">
      <div v-if="$slots.inputs">
        <slot name="inputs" />
      </div>

      <div class="row justify-end">
        <q-btn
          :loading="requestPending"
          rounded
          no-caps
          color="primary"
          size="md"
          class="col-12 col-sm-auto"
          style="min-width: 180px"
          @click="emit('search')"
        >
          {{ searchLabel }}

          <template v-slot:loading>
            <q-spinner-hourglass class="on-left" />
            {{ $t('hotel.search.request_pending') }}
          </template>
        </q-btn>
      </div>
    </q-card-section>
  </q-card>
</template>

<script lang="ts" setup>
defineProps<{
  requestPending: boolean
  searchLabel: string
}>()

const emit = defineEmits<{ search: unknown }>()
</script>
