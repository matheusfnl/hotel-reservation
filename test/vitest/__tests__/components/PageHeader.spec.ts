import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { QBtn } from 'quasar'

import PageHeader from '@/components/PageHeader.vue'

const mountWrapper = (props: {
  requestPending: boolean
  searchLabel: string
  disabled: boolean
}) => {
  return mount(PageHeader, {
    props,
  })
}

describe('PageHeader', () => {
  it('should render correctly', async () => {
    const wrapper = mountWrapper({ requestPending: false, searchLabel: 'Search', disabled: false })
    const button = wrapper.findComponent(QBtn)

    expect(wrapper.html()).toMatchSnapshot()
    expect(button.props('loading')).toBe(false)
    expect(button.attributes('disabled')).not.toBeDefined()

    await button.trigger('click')

    expect(wrapper.emitted()).toHaveProperty('search')
  })

  it('should disable button when disabled prop is true', () => {
    const wrapper = mountWrapper({ requestPending: false, searchLabel: 'Search', disabled: true })
    const button = wrapper.findComponent(QBtn)

    expect(button.attributes('disabled')).toBeDefined()
  })

  it('should show loading state when requestPending is true', () => {
    const wrapper = mountWrapper({ requestPending: true, searchLabel: 'Search', disabled: false })
    const button = wrapper.findComponent(QBtn)

    expect(button.props('loading')).toBe(true)
    expect(button.text()).toContain(wrapper.vm.$t('hotel.search.request_pending'))
  })

  it('should emit search event on button click', async () => {
    const wrapper = mountWrapper({ requestPending: false, searchLabel: 'Search', disabled: false })
    const button = wrapper.findComponent(QBtn)

    await button.trigger('click')

    expect(wrapper.emitted()).toHaveProperty('search')
  })
})
