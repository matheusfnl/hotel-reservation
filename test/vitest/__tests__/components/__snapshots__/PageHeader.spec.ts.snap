// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageHeader > should render correctly 1`] = `
"<div class="q-card bg-white rounded-borders-md">
  <div class="q-card__section q-card__section--vert q-px-md q-py-sm">
    <h2 class="text-h5 q-ma-none">Reservar hotel</h2>
  </div>
  <hr class="q-separator q-separator--horizontal" aria-orientation="horizontal">
  <div class="q-card__section q-card__section--vert column q-gutter-md">
    <!--v-if-->
    <div class="row justify-end"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase col-12 col-sm-auto" style="font-size: 14px; min-width: 180px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row">Search </span>
        <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
          <!---->
        </transition-stub>
      </button></div>
  </div>
</div>"
`;
